### ISSUE TITLE
Update to 3.3.0 issue

### ISSUE THREAD

**Issue by axelex** [ceolter/ag-grid#754]
I've updated to version 3.3.0 from 2.3.5 and I get this error 'Uncaught ReferenceError: exports is not defined' 
Is there anything extra I should do to make the new version work or do you know why I'm encountering this issue?

Thanks!

**Comment by leocaseiro** [comment]
:+1: Same for me!

Angular 1.5.0
Trying to create an Angular module and register the ag-grid Directive:
```
agGrid.initialiseAgGridWithAngular1(angular);
```

I believe is something with to use webpack or not use.
Ref: https://www.ag-grid.com/best-angularjs-data-grid/index.php

**Comment by leocaseiro** [comment]
I'm loading all my third-party libraries from their bower.json. I just load the ```main``` file.
I realised that the main file in bower has been replaced on commit f014f714741263c2:

From

```
"main": [
     "dist/ag-grid.js",
     "dist/ag-grid.css",		
     "dist/theme-fresh.css",		
     "dist/theme-blue.css",		
     "dist/theme-dark.css"		
],
````

To

```
"main": [
    "./main.js"
 ],
```

I replaced on my workflow to the file ```dist/ag-grid.js``` instead of ```main.js``` and now it's working properly. Hope works for you.

**Comment by ceolter** [comment]
not enough info to help. exports is what commonjs files use. so it sounds like your referencing the ag-Grid commonjs files without actually using commonjs.

if using commonjs, use the main.js in the root.

if not using commonjs, using the files in dist/

**Comment by axelex** [comment]
We are using gulp to generate a vendor.js file that contains all js and css files in bower.
The issue was that because of this line in .bower.json '"main": [
    "./main.js"
  ],', our gulp task loads only main.js. I changed that to "main": [
    "dist/ag-grid.js",
    "dist/ag-grid.css"
  ], and it works now, but I also had to add this line 'agGrid.initialiseAgGridWithAngular1(angular);' before injecting the 'agGrid' dependency with this update which even at the 3.2.0 wasn't needed. 
I know this is because it works on different platforms and you need to somehow initialize it depending on framework but still is an extra change that need to be made so we can keep an updated version.

Aniway changing the main section of .bower.json is most important issue right now for us because it would mean changing it every time we do a bower install... do you have any sugestions of avoiding that?

**Comment by leocaseiro** [comment]
Hey @axelex, keep your `bower.json` the way @ceolter left on bower and fix your stream on gulp.

You can use [merge2](https://www.npmjs.com/package/merge2) combined to [gulp-filter](https://www.npmjs.com/package/gulp-filter) to load the correct files, like so:

```javascript
var gulpFilter = require('gulp-filter');
var merge2     = require('merge2'),

gulp.task('default', function () {
    return gulp.src('src/**/*.js')

    	//merge 2 streams in one
    	merge2(
    		gulp.src(automaticFilesFromMainBowert),
    		gulp.src([
    			'./bower_components/ag-grid/dist/ag-grid.js',
    			'./bower_components/ag-grid/dist/ag-grid.css',
    		])
		)
        //Remove the main.js from your stream
        .pipe(gulpFilter(['**', '!*bower_components/ag-grid/main.js']))
        
        //...Do things you need

        .pipe(gulp.dest());
});
```

**Comment by axelex** [comment]
Thanks a lot Leo! 
I will try that tomorrow. I'll let you know how that worked.

**Comment by leocaseiro** [comment]
:beers: No worries! Hope works for you!

**Comment by axelex** [comment]
It's adding the ag-grid css and js files now but for some reason it's not filtering the main.js... I tried different ways to filter that but it's not filtered...

**Comment by leocaseiro** [comment]
Hi @axelex you have to make sure you use the exclamation mark, like my example:

```javascript
    //Remove the main.js from your stream
    .pipe(gulpFilter(['**', '!*/ag-grid/main.js']))
```

**Comment by axelex** [comment]
Yes I did... but it doesn't exclude the file...

**Comment by leocaseiro** [comment]
can you post your gulp task here please?

**Comment by axelex** [comment]
Yes.

Here it is as it is now, were I remove the file manually with js:
new Task('bower-js', function() {
        var filesToInclude = [];
        mainBowerFiles({includeDev:true}).forEach(function(file) {
            if(!(file.indexOf('bower_components/ag-grid/main.js') > -1))
            {
                filesToInclude.push(file);
            }
        });

        return merge2(
                gulp.src(filesToInclude),
                gulp.src(['./bower_components/ag-grid/dist/ag-grid.js'])
            )
            .on('error', onError)
            .pipe(filter('**/*.js'))
            .pipe(concat(jsFile, {sourcesContent: true}))
            .pipe(gulpIf(Elixir.config.production, uglify()))
            .pipe(gulp.dest(jsOutputFolder || Elixir.config.js.outputFolder))
            .pipe(notify({
                title: 'Laravel Elixir',
                subtitle: 'Javascript Bower Files Imported!',
                icon: __dirname + '/../node_modules/laravel-elixir/icons/laravel.png',
                message: ' '
            }));
    }).watch('bower.json');


and these are a few of the methods I tried:

1.
 )
            .on('error', onError)
            .pipe(filter(['**/*.js',  '!*bower_components/ag-grid/main.js']))
            .pipe(concat(jsFile, {sourcesContent: true}))
2.
 )
            .on('error', onError)
            .pipe(filter('**/*.js'))
            .pipe(filter(['**',  '!*bower_components/ag-grid/main.js']))
            .pipe(concat(jsFile, {sourcesContent: true}))
3.
 )
            .on('error', onError)
            .pipe(filter('**/*.js'))
            .pipe(filter('!*bower_components/ag-grid/main.js'))
            .pipe(concat(jsFile, {sourcesContent: true}))

**Comment by leocaseiro** [comment]
I'm not sure why you have an `.on('error')`, but you suppose to filter just after your mege2, like so.
Also, you need to leave **

```javascript
return merge2(
            gulp.src(filesToInclude),
            gulp.src(['./bower_components/ag-grid/dist/ag-grid.js'])
        )
        .pipe(filter(['**', '!*/ag-grid/main.js']))
        .on('error', onError)
        .pipe(filter('**/*.js')) //I also think you don't need this, only if you want to ignore the .css files

**Comment by axelex** [comment]
I have separate tasks for js and css, that's why I added a filter only on js files and I figured the logical thing would be to just add the exception after that... I've researched a bit about filtering main bower files and it should work as I tried before... I will try tomorrow though to filter right after merging and comment the `on('error')` line to see what happens...

**Comment by leocaseiro** [comment]
Use [gulp-debug](https://www.npmjs.com/package/gulp-debug) to check the files you have in your stream as well:

```javascript

return merge2(
        gulp.src(filesToInclude),
        gulp.src(['./bower_components/ag-grid/dist/ag-grid.js'])
    )
        .pipe(debug({title: 'before remove filter'}))
        .pipe(filter(['**', '!*/ag-grid/main.js'])) //don't change this line
        .pipe(debug({title: 'after remove filter'}))
        .pipe(filter('**/*.js'))
        .pipe(debug({title: 'after second .js filter'}))
```

I would get rid of .on('error') for testing first. I'm not sure if that's causing an issue...

**Comment by axelex** [comment]
I did that but it's ignoring the main.js filter:
`[12:28:12] after second .js filter bower_components/ngstorage/ngStorage.js
[12:28:12] before remove filter bower_components/angular-mocks/angular-mocks.js
[12:28:12] after remove filter bower_components/angular-mocks/angular-mocks.js
[12:28:12] after second .js filter bower_components/angular-mocks/angular-mocks.js
[12:28:12] before remove filter bower_components/ag-grid/main.js
[12:28:12] after remove filter bower_components/ag-grid/main.js
[12:28:12] after second .js filter bower_components/ag-grid/main.js
[12:28:12] before remove filter bower_components/angular-chart.js/dist/angular-chart.js
[12:28:12] after remove filter bower_components/angular-chart.js/dist/angular-chart.js`

But it recognizes the second filter:

`[12:28:12] before remove filter 87 items
[12:28:12] after remove filter 87 items
[12:28:12] after second .js filter 52 items`

**Comment by leocaseiro** [comment]
Not nice!

It seems they are all happening in parallel...maybe, we should try one by one and see how it works.

Also I found this fix for mainBowerFiles: 
https://github.com/ck86/main-bower-files/issues/117

It seems you're able to do your filter only js and also ignoring the main.js from ag-grid only with mainBowerFiles. 

Perhaps like that:
```javascript
mainBowerFiles(['**/*.js', '!**/ag-grid/main.js'], {includeDev:true});
```

**Comment by leocaseiro** [comment]
Btw, the first filter ` gulp.src(['./bower_components/ag-grid/dist/ag-grid.js'])` is not working.

Where is your bower_components folder? Is at the same root as your gulpfile.js or behind it?
Because this interfere on your filters like `../` or `./`

```
Project
+-- /bower_components
+-- gulpfile.js

Alternative 2
+-- /bower_components
    +- /Dev
       +-- gulpfile.js
```

**Comment by axelex** [comment]
They are on the same level, and the ag-grid.js is loaded properly so that statement is working. the only thing that wasn't working was the filter for main.js. 
But now after doing the filtering directly from mainBowerFiles that works as well.

Thanks a lot @leocaseiro!
You've helped me a lot and you've been really patient!

**Comment by leocaseiro** [comment]
No worries @axelex . I'm glad worked it!

You know, developers are patient in general!

:beers:

**Comment by ceolter** [comment]
gents, thanks for helping each other out. pretty busy my end to look into requests that are more than simple one liner questions.

at the end of it all, do i need to add anything to my documentation or change the grid code?

**Comment by leocaseiro** [comment]
@ceolter, if you must keep `{main: ['./main.js']}` instead of the previous, so I believe would be nice having a Warning for who is using **gulp/grunt** with [mainBowerFiles](https://github.com/ck86/main-bower-files), [wiredep](https://www.npmjs.com/package/wiredep), or any script that read files from bower.json `main`.

Warning that they need to remove main.js and replace with dist/

**Comment by axelex** [comment]
Yes, that would be great!

**Comment by ceolter** [comment]
is this basically a difference between bundled vs commonjs versions of the grid?

so maybe i add to the docs:
```
warning: if you are using the bundled version of the grid (/dist/ag-grid-js) then you must specify this directly, the main files specified in package.json and bower.json point to the commonjs versions of ag-Grid
```

would that do it?

**Comment by leocaseiro** [comment]
LGTM

**Comment by germanger** [comment]
I updated from 3.2.0 to 3.3.3 and I'm using angular 1.
If this helps to someone, I had to change the following things:

Before my angular app definition, ie: `angular.module('myApp', [..., agGrid, ...])`, I had to add this line:

          agGrid.initialiseAgGridWithAngular1(angular); // <---
          angular.module('myApp', [..., agGrid, ...])

Also, I use Grunt's `bower-concat` to concatenate all my third party libraries into two files:  `bower-concat.js` and `bower-concat.css`

This configuration worked for `bower-concat`:

    bower_concat: {
        all: {
            dest: 'build/bower-concat.js',
            cssDest: 'build/bower-concat.css',
            mainFiles: {
                'agGrid': ['dist/ag-grid.js', 'dist/ag-grid.css']     // I had to manually specify the main files
            },
            bowerOptions: {
                relative: false
            }
        }
    },

Hope it helps someone.

**Comment by tanou73** [comment]
Hi lads !
Just a tip for those who use wiredeps and face the same issue, I think the must simple way to go is to use this very handy feature of wiredep: https://github.com/taptapship/wiredep#bower-overrides.

This allows you to override the main package of any of your dependencies !
